.github/actions/perform-deployment/index.js

```diff
@@ -0,0 +1,123 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+const core = require('@actions/core');
+const cache = require('@actions/cache');
+const fs = require('fs');
+const path = require('path')
+const child_process = require('child_process')
+const buffer = require('buffer')
+const readline = require('readline')
+
+
+async function run() {
+  try {
+    const deploymentName = core.getInput("deployment-name")
+    const cacheVersionSuffix = core.getInput("cache-version-suffix")
+    const cacheDir = "deployment_caches"
+
+    // Get json list with names of all deployments which are needed for this deployment.
+    const deployment_helper_script_path = path.join("agent_build" ,"scripts", "run_deployment.py")
+    // Run special github-related helper command which returns names for all deployments, which are used in the current
+    // deployment.
+    const code = child_process.execFileSync(
+        "python3",
+        [deployment_helper_script_path, "deployment", deploymentName, "get-deployment-all-cache-names"]
+    );
+
+    // Read and decode names from json.
+    const json_encoded_deployment_names = buffer.Buffer.from(code, 'utf8').toString()
+    const deployer_cache_names = JSON.parse(json_encoded_deployment_names)
+
+    const cache_hits = {}
+
+    // Run through deployment names and look if the is any existing cache for them.
+    for (let name of deployer_cache_names) {
+
+        const cache_path = path.join(cacheDir, name)
+        const key = `${name}-${cacheVersionSuffix}`
+
+        // try to restore the cache.
+        const result = await cache.restoreCache([cache_path], key)
+
+        if(typeof result !== "undefined") {
+          console.log(`Cache for the deployment ${name} is found.`)
+        } else {
+          console.log(`Cache for the deployment ${name} is not found.`)
+        }
+        cache_hits[name] = result
+
+    }
+
+    // Run the deployment. Also provide cache directory, if there are some found caches, then the deployer
+    // has to reuse them.
+    child_process.execFileSync(
+        "python3",
+        [deployment_helper_script_path, "deployment", deploymentName, "deploy", "--cache-dir", cacheDir],
+        {stdio: 'inherit'}
+    );
+
+    if ( fs.existsSync(cacheDir)) {
+      console.log("Cache directory is found.")
+
+      // Run through the cache folder and save any cached directory within, that is not yet cached.
+      const filenames = fs.readdirSync(cacheDir);
+      for (const name of filenames) {
```

Comment:
I wonder if we could refactor this code path a bit to reduce levels of indentation and make it a bit easier to read.

Perhaps return early in case cacheDir doesn't exist ?